### PR TITLE
[pytest][Multi ASIC] Add multi ASIC test for QoS DSCP verification on backend ASICs

### PR DESCRIPTION
**Pull request by smaheshm**
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
Added multi ASIC test for QoS DSCP verification on backend ASICs. Test is not applicable to single ASIC platforms.

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Approach
#### What is the motivation for this PR?
New QoS DSCP test case to verify on backend ASICs on multi ASIC platform

#### How did you do it?
The majority of the code changes is to identify the tx and rx ports on a backend ASIC.

Added parameterization for backend ASICs so the test runs for each backend ASIC. Following are the steps to identify tx and rx ports.

* Find the source and destination ASIC, these should be separate ASICs so that traffic gets forwarded by backend ASICs.
* On the source ASIC shutdown the backend ASIC IFs to all backedn ASICs except the one under test. This ensures traffic is forwarded by the ASIC under test.
* Send a burst of packets to identify the tx and rx ports on the backend ASIC under test.
* Pass these ports to saitest to perform DSCP test.

#### How did you verify/test it?

```
samaddik@svcstr-server-2:/var/sonic-mgmt/tests$ pytest --count=2 qos/test_qos_sai.py  --testbed=vmsvc1-t1-n3164-acs-1 --inventory=../ansible/strsvc  --testbed_file=../ansible/testbed.csv --host-pattern=svcstr-n3164-acs-1  --module-path=../ansible/library --disable_loganalyzer --skip_sanity  -k masic_dscp_queue
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
========================================================================================================================================================================== test session starts ===========================================================================================================================================================================
platform linux2 -- Python 2.7.17, pytest-4.6.5, py-1.10.0, pluggy-0.13.1
ansible: 2.8.12
rootdir: /var/sonic-mgmt/tests, inifile: pytest.ini
plugins: html-1.22.1, repeat-0.9.1, forked-1.3.0, xdist-1.28.0, metadata-1.11.0, ansible-2.2.2
collected 156 items / 152 deselected / 4 selected                                                                                                                                                                                                                              

qos/test_qos_sai.py ....                                                                                                                                                                                                                                                                                                                                           [100%]

============================================================================================================================================================================ warnings summary ============================================================================================================================================================================
/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:538
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:538: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.dualtor
    self.import_plugin(import_spec)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========================================================================================================================================================= 4 passed, 152 deselected, 1 warnings in 464.87 seconds =========================================================================================================================================================

```

#### Any platform specific information?
Relevant to Multi ASIC platforms

#### Supported testbed topology if it's a new test case?

### Documentation 
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

### CODE REVIEW

File: tests/common/devices/sonic.py [hunk 1]
```diff
@@ -1340,3 +1340,69 @@ def get_up_ip_ports(self):
             except KeyError:
                 pass
         return up_ip_ports
+
+    def get_rsyslog_ipv4(self):
```
Comment (neethajohn) [inline tests/common/devices/sonic.py:1344]: These changes should already be in through other PRs. Can you update the branch and push?

Comment (smaheshm) [reply tests/common/devices/sonic.py:1344]: that' right, that's why I didn't assign anyone yet.. It also needs a merge.

**Comment by smaheshm** [comment]
🛎️

File: tests/qos/qos_sai_base.py [hunk 2]
```diff
@@ -1070,3 +1053,295 @@ def resetWatermark(
...
+                .
+            }
+           .
+           .
+        }
+        """
+        duthost = duthosts[rand_one_dut_hostname]
+
+        topo = tbinfo["topo"]["name"]
+        if topo not in self.SUPPORTED_T1_TOPOS:
```
Comment (arlakshm) [inline tests/qos/qos_sai_base.py:1270]: Is this fixture used only for multi asic platform ?

File: tests/qos/qos_sai_base.py [hunk 3]
```diff
@@ -1070,3 +1053,295 @@ def resetWatermark(
...
+        def is_intf_status(asic, intf, oper_state):
+            intf_status = duthost.asic_instance(asic).show_interface(
+                command="status", include_internal_intfs=True
+            )["ansible_facts"]["int_status"]
+            if intf_status[intf]["oper_state"] == oper_state:
+                return True
+            return False
+
+        asic_ns = "asic{}".format(test_asic)
+        cmd_str = "sudo config interface -n asic{} {} {}"
```
Comment (arlakshm) [inline tests/qos/qos_sai_base.py:1142]: use this function in sonicAsic https://github.com/Azure/sonic-mgmt/blob/a5ea4bf6b77438dfe1e6d3f413c055e7c04470c2/tests/common/devices/sonic_asic.py#L448  iso using the cli directly ?

File: tests/saitests/sai_qos_tests.py [hunk 4]
```diff
@@ -235,28 +252,34 @@ def runTest(self):
...
-                dscp_received = False
-                while not dscp_received:
-                    result = self.dataplane.poll(device_number=0, port_number=dst_port_id, timeout=3)
-                    if isinstance(result, self.dataplane.PollFailure):
-                        self.fail("Expected packet was not received on port %d. Total received: %d.\n%s" % (dst_port_id, cnt, result.format()))
-                    recv_pkt = scapy.Ether(result.packet)
-                    cnt += 1
+                # skip this check for backend ASIC, rely on queue counters
+                # TTL changes on multi ASIC platforms
+                if test_dst_port_name is None:
```
Comment (arlakshm) [inline tests/saitests/sai_qos_tests.py:257]: instead of bypassing this check for multi asic, we can skip checking testing the ttl ?

**Review by arlakshm — COMMENTED** [review]
Few minor comments, but change lgtm

File: tests/qos/qos_sai_base.py (hunk 2 above)
Comment (smaheshm) [reply tests/qos/qos_sai_base.py:1270]: yes, may be there should be a fixture to bail out when not on MASIC platform.

File: tests/saitests/sai_qos_tests.py (hunk 4 above)
Comment (smaheshm) [reply tests/saitests/sai_qos_tests.py:257]: It's fine with me, @neethajohn Fine to remove TTL check? The exp_ip_id alone is enough to guarantee it's the intended packet.

File: ansible/library/show_interface.py [hunk 5]
```diff
@@ -194,11 +194,15 @@ def collect_interface_status(self, namespace=None, include_internal_intfs=False)
 
         return
 
-    def collect_interface_counter(self):
+    def collect_interface_counter(self, namespace=None, include_internal_intfs=False):
         regex_int = re.compile(r'\s*(\S+)\s+(\w)\s+([,\d]+)\s+(N\/A|[.0-9]+ B/s)\s+(\S+)\s+([,\d]+)\s+(\S+)\s+([,\d]+)\s+([,\d]+)\s+(N\/A|[.0-9]+ B/s)\s+(\S+)\s+([,\d]+)\s+(\S+)\s+([,\d]+)')
         self.int_counter = {}
+        cli_options = " -n {}".format(namespace) if namespace is not None else ""
+        if include_internal_intfs:
```
Comment (neethajohn) [inline ansible/library/show_interface.py:201]: can you also check that namespace exists? similar to the check in collect_interface_status to enable pytest runs on 201811 images

File: tests/saitests/sai_qos_tests.py [hunk 6]
```diff
@@ -2114,3 +2140,45 @@ def runTest(self):
 
         finally:
             sai_thrift_port_tx_enable(self.client, asic_type, [dst_port_id])
+
+
+class PacketTransmit(sai_base_test.ThriftInterfaceDataPlane):
```
Comment (neethajohn) [inline tests/saitests/sai_qos_tests.py:2145]: where is this used?

File: tests/qos/qos_sai_base.py [hunk 7]
```diff
@@ -44,6 +39,87 @@ def isBufferInApplDb(self, dut_asic):
...
+class QosSaiBase(QosBase):
+    """
+        QosSaiBase contains collection of pytest fixtures that ready the tesbed for QoS SAI test cases.
+    """
+    SUPPORTED_ASIC_LIST = ["td2", "th", "th2", "spc1", "spc2", "spc3", "td3"]
+    TARGET_QUEUE_WRED = 3
+    TARGET_LOSSY_QUEUE_SCHED = 0
+    TARGET_LOSSLESS_QUEUE_SCHED = 3
+    buffer_model_initialized = False
+    buffer_model = None
```
Comment (neethajohn) [inline tests/qos/qos_sai_base.py:121]: These should be initialized in the base clase

**Review by neethajohn — APPROVED** [review]

**Pull request merged**